### PR TITLE
feat(timestream): Adds L2 constructs for Timestream

### PR DESCRIPTION
**Pull request by kochie**
This PR adds L2 constructs for all current timestream objects
- `Database`
- `Table`
- `ScheduledQuery`

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
	* [x] Did you use `cdk-integ` to deploy the infrastructure and generate the snapshot (i.e. `cdk-integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by kochie** [comment]
I've just realized that all new L2 constructs are in separate alpha packages. Would this be better in one of those?

File: packages/@aws-cdk/aws-timestream/package.json [hunk 1]
```diff
@@ -81,25 +81,38 @@
...
+    "@aws-cdk/aws-sns": "0.0.0",
     "@aws-cdk/core": "0.0.0",
     "constructs": "^10.0.0"
   },
   "engines": {
     "node": ">= 14.15.0"
   },
-  "stability": "experimental",
-  "maturity": "cfn-only",
+  "stability": "stable",
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-timestream/package.json:114]: Please don't declare this to be stable immediately. It should start out as `experimental/experimental`.

File: packages/@aws-cdk/aws-timestream/lib/scheduled-query.ts [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+   */
+  readonly scheduleConfiguration: ScheduleConfiguration
+
+  /**
+   * The ARN for the IAM role that Timestream will assume when running the scheduled query.
+   */
+  readonly scheduledQueryExecutionRole: IRole
+
+  /**
+   * A name for the query. Scheduled query names must be unique within each Region.
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-timestream/lib/scheduled-query.ts:404]: Account and region, I hope?

Can we autogenerate this?

File: packages/@aws-cdk/aws-timestream/lib/table.ts [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+   * The encryption option for the S3 location. Valid values are S3 server-side encryption with an S3 managed key (SSE_S3) or AWS managed key (SSE_KMS).
+   */
+  readonly encryptionOption: EncryptionOptions,
+
+  /**
+   * The AWS KMS key ID to use when encrypting with an AWS managed key.
+   *
+   * @default None
+   */
+  readonly kmsKeyId?: string,
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-timestream/lib/table.ts:51]: `IKey`

File: packages/@aws-cdk/aws-timestream/lib/table.ts [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+export interface MagneticS3Configuration {
+  /**
+   * The name of the S3 bucket.
+   */
+  readonly bucketName: string,
+
+  /**
+   * The encryption option for the S3 location. Valid values are S3 server-side encryption with an S3 managed key (SSE_S3) or AWS managed key (SSE_KMS).
+   */
+  readonly encryptionOption: EncryptionOptions,
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-timestream/lib/table.ts:44]: Should this be optional?

File: packages/@aws-cdk/aws-timestream/lib/table.ts [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+/**
+ * Contains properties to set on the table when enabling magnetic store writes.
+ */
+export interface MagneticStoreWriteProperties {
+  /**
+   * A boolean flag to enable magnetic store writes.
+   */
+  readonly enableMagneticStoreWrites: boolean
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-timestream/lib/table.ts:78]: Could be optional and default to true?

Also could be called `enabled`?

File: packages/@aws-cdk/aws-timestream/lib/table.ts [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+   * Contains properties to set on the table when enabling magnetic store writes.
+   *
+   * @default None
+   */
+  readonly magneticStoreWriteProperties?: MagneticStoreWriteProperties,
+
+  /**
+   * The retention duration for the memory store and magnetic store. This object has the following attributes:
+   *
+   * @default None
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-timestream/lib/table.ts:122]: Describe what happens.

**Review by rix0rrr — CHANGES_REQUESTED** [review]
The property hierarchy is very nested, and many things are required (do they need to be? Can we have sensible defaults?)

We prefer property structures to be flatter, and have as few required properties as possible:

https://github.com/aws/aws-cdk/blob/master/docs/DESIGN_GUIDELINES.md#flat

File: packages/@aws-cdk/aws-timestream/lib/table.ts (hunk 4 above)
Comment (LukvonStrom) [reply packages/@aws-cdk/aws-timestream/lib/table.ts:44]: What would a sensible default look like here?

File: packages/@aws-cdk/aws-timestream/lib/database.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
+import { UnknownPrincipal } from '@aws-cdk/aws-iam';
+import { IKey } from '@aws-cdk/aws-kms';
+import { ArnFormat, IResource, Resource, Stack, Token } from '@aws-cdk/core';
+import { Construct } from 'constructs';
+import { CfnDatabase } from './timestream.generated';
+
+/**
+ * Timestream Database interface
+ */
+export interface IDatabase extends IResource {
```
Comment (LukvonStrom) [inline packages/@aws-cdk/aws-timestream/lib/database.ts:10]: Could you implement `IGrantable` as well? Makes it more ergonomic

File: packages/@aws-cdk/aws-timestream/lib/database.ts [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    this.databaseArn = this.getResourceArnAttribute(resource.attrArn, {
+      service: 'timestream',
+      resource: this.physicalName,
+    });
+    this.databaseName = this.getResourceNameAttribute(resource.ref);
+  }
+
+  private validateDatabaseName(databaseName: string) {
+    if (!Token.isUnresolved(databaseName)) {
+      const databaseNameLength = Buffer.byteLength(databaseName, 'utf-8');
```
Comment (LukvonStrom) [inline packages/@aws-cdk/aws-timestream/lib/database.ts:116]: why are you not using string length?

**Comment by LukvonStrom** [comment]
@kochie Implemented some of the proposed changes in https://github.com/kochie/aws-cdk-1/pull/14, leveraging your previous work :)

**Comment by kochie** [comment]
> @kochie Implemented some of the proposed changes in [kochie#14](https://github.com/kochie/aws-cdk-1/pull/14), leveraging your previous work :)

Thanks @LukvonStrom, added changes to branch!

File: packages/@aws-cdk/aws-timestream/lib/table.ts (hunk 4 above)
Comment (kochie) [reply packages/@aws-cdk/aws-timestream/lib/table.ts:61]: I'm thinking if this is going to be set a default SSE_S3 seems reasonable?

**Comment by LukvonStrom** [comment]
Fixed the build + implemented a more flat API surface in the kochie#16 PR

**Comment by LukvonStrom** [comment]
@peterwoodworth Why is the build failing here? What exactly should be in the README for the stability banner? Is there a documentation on this?

**Comment by peterwoodworth** [comment]
I'm not sure we have documentation on the stability banner since I believe it's automatically generated when the README is created for the new L1 service offered by CFN. Take a look in this PR and you'll see that you're actually removing the existing stability banner for this module which indicates that all Cfn resources are stable

Since you're creating new L2 constructs, you'll want to indicate that these new constructs are unstable. Check the[ readme of a module still in experimental state](https://raw.githubusercontent.com/aws/aws-cdk/master/packages/%40aws-cdk/aws-apigatewayv2/README.md) to see how to do this 🙂

**Comment by LukvonStrom** [comment]
@kochie I guess we need guidance from @peterwoodworth ?

**Comment by skinny85** [comment]
I think running `yarn pkglint` in the directory of the module should populate the ReadMe for you.

**Comment by TheRealAmazonKendra** [comment]
This PR has been deemed to be abandoned, and will be automatically closed. Please create a new PR for these changes if you think this decision has been made in error.

**Pull request closed without merging**